File: binary/python3/path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if root is None:
            return False
        
        sum -= root.val
        
        if sum == 0 and root.left is None and root.right is None:
            return True
        
        if sum != 0 and root.left is None and root.right is None:
            return False
        
        hasPath = False
        if root.left is not None:
            hasPath = hasPath or self.hasPathSum(root.left, sum)
        
        if root.right is not None:
            hasPath = hasPath or self.hasPathSum(root.right, sum)
            
        return hasPath

File: binary/cpp/path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (!root) {
            return false;
        }
        if (!root->left && !root->right) {
            return sum == root->val;
        }
        int restVal = sum - root->val;
        bool leftAnswer, rightAnswer;
        leftAnswer = rightAnswer = false;
        if (root->left) {
            leftAnswer = hasPathSum(root->left, restVal);
        }
        if (root->right) {
            rightAnswer = hasPathSum(root->right, restVal);
        }
        return (leftAnswer || rightAnswer);
    }
};